File: rag/src/retriever.py
"""
retriever.py - 混合检索器。

组合 SQLite FTS、可选向量检索、轻量重排和 source_type 权重，
输出最终 top-k chunks。
"""

from . import router as router_mod
from .embedder import embed_text
from .storage import sqlite_store
from .storage import vector_store

SOURCE_WEIGHTS = {"rule": 1.0, "knowledge": 0.9, "reference": 0.8, "project": 0.6}


def hybrid_retrieve(query, task_type=None, top_k=5, fts_n=15, vec_n=15):
    """执行混合检索，并返回重排后的 top-k chunks。"""
    import time

    start_time = time.time()
    query_lower = query.lower()

    if task_type:
        route_result = {"task_type": task_type, **router_mod.get_route(task_type)}
    else:
        route_result = router_mod.route_query(query)
    task_route = route_result

    fts_results = []
    categories = task_route.get("categories", [])
    stages = task_route.get("stages", [])

    if categories:
        filtered = sqlite_store.filter_chunks(categories=categories, stages=stages, top_n=fts_n)
        for row in filtered:
            text = (row.get("title", "") + " " + row.get("text", "")).lower()
            terms = [term for term in query_lower.split() if term]
            score = 0
            for term in terms:
                if term in text:
                    score -= 1

            title_text = row.get("title", "").lower()
            if any(term in title_text for term in terms):
                score -= 3

            row["fts_score"] = score if score < 0 else -0.1
            fts_results.append(row)

        fts_results = [r for r in fts_results if r.get("fts_score", 0) < 0]

    if not fts_results:
        fts_results = sqlite_store.fts_search(query, fts_n)

    fts_map = {row["chunk_id"]: row for row in fts_results}

    vec_results = []
    try:
        query_vector = embed_text(query)
        vec_results = vector_store.vector_search(query_vector, vec_n)
    except Exception as exc:
        print(f"[retriever] 向量检索失败: {exc}")

    vec_map = {row["chunk_id"]: row["score"] for row in vec_results}

    candidates = {}
    for chunk_id, info in fts_map.items():
        candidates[chunk_id] = {
            "chunk_id": chunk_id,
            "row": info,
            "fts_score": info.get("fts_score"),
            "vec_score": vec_map.get(chunk_id),
        }

    for row in vec_results:
        if row["chunk_id"] in candidates:
            continue

        chunk_data = sqlite_store.get_chunk_by_id(row["chunk_id"])
        if chunk_data:
            candidates[row["chunk_id"]] = {
                "chunk_id": row["chunk_id"],
                "row": chunk_data,
                "fts_score": None,
                "vec_score": row["score"],
            }

    reranked = []
    for chunk_id, candidate in candidates.items():
        row = candidate["row"]
        score = _rerank(row, query_lower, task_route, candidate["fts_score"], candidate["vec_score"])
        reason = _build_reason(row, candidate["fts_score"], candidate["vec_score"], task_route)
        text = row.get("text", "")
        reranked.append({
            "chunk_id": chunk_id,
            "title": row.get("title", ""),
            "score": score,
            "reason": reason,
            "snippet": text[:200] + ("..." if len(text) > 200 else ""),
            "source_path": row.get("source_path", ""),
            "source_type": row.get("source_type", ""),
            "category": row.get("category", ""),
            "tags": row.get("tags", []),
            "priority": row.get("priority", 3),
        })

    reranked.sort(key=lambda x: x["score"], reverse=True)
    elapsed = time.time() - start_time

    return {
        "results": reranked[:top_k],
        "meta": {
            "total_candidates": len(candidates),
            "fts_count": len(fts_results),
            "vector_count": len(vec_results),
            "elapsed_ms": round(elapsed * 1000),
            "task_type": route_result.get("task_type"),
            "confidence": route_result.get("confidence", 0),
        },
    }


def _rerank(row, query_lower, task_route, fts_score, vector_score):
    if isinstance(fts_score, (int, float)) and fts_score < 0:
        fts_norm = min(1.0, max(0.0, -fts_score / 15))
    else:
        fts_norm = 0.0

    if isinstance(vector_score, (int, float)):
        vec_norm = min(1.0, max(0.0, vector_score))
    else:
        vec_norm = 0.2

    task_match = 0.0
    if task_route:
        categories = task_route.get("categories", [])
        if categories and row.get("category") in categories:
            task_match += 0.5

        stages = task_route.get("stages", [])
        if row.get("stage") and any(stage in str(row["stage"]) for stage in stages):
            task_match += 0.3

        description = (task_route.get("description", "") or "").lower()
        query_terms = [term for term in query_lower.split() if len(term) > 1]
        import re

        desc_terms = re.split(r"[：:、，。\s]", description)
        overlap = sum(1 for qt in query_terms if any(qt in rt for rt in desc_terms))
        task_match += min(overlap * 0.05, 0.2)

    task_match = min(1.0, task_match)

    priority_weight = (6 - (row.get("priority", 3) or 3)) / 5
    source_weight = SOURCE_WEIGHTS.get(row.get("source_type"), 0.5)

    return round(
        vec_norm * 0.35
        + fts_norm * 0.15
        + task_match * 0.30
        + priority_weight * 0.12
        + source_weight * 0.08,
        4,
    )


def _build_reason(row, fts_score, vector_score, task_route):
    parts = []
    if isinstance(fts_score, (int, float)) and fts_score < 0:
        parts.append("关键词匹配")
    if isinstance(vector_score, (int, float)) and vector_score > 0.3:
        parts.append("语义接近")
    if task_route and row.get("category") in task_route.get("categories", []):
        parts.append("任务类型命中")
    if row.get("priority", 3) <= 2:
        parts.append("高优先级")
    if row.get("source_type") == "rule":
        parts.append("规则文档")
    if row.get("source_type") == "knowledge":
        parts.append("知识包")
    if not parts:
        parts.append("综合匹配")
    return " + ".join(parts)

File: rag/src/retriever.js
/**
 * retriever.js — 混合检索器
 *
 * 职责：
 * - 混合 FTS + 向量检索
 * - 轻量重排器
 * - 输出最终 top-k chunks
 *
 * 重排公式：
 *   最终分 = 语义分 * 0.35 + FTS分 * 0.15 + 任务匹配分 * 0.30 + priority权重 * 0.12 + source_type权重 * 0.08
 */

import { readFileSync, existsSync, mkdirSync } from 'fs';
import { join, resolve, dirname } from 'path';
import { fileURLToPath } from 'url';
import { createRequire } from 'module';
import * as sqliteStore from './storage/sqlite-store.js';
import * as vectorStore from './storage/vector-store.js';
import * as router from './router.js';
import { embedText } from './embedder.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = resolve(__dirname, '../..');
const require = createRequire(import.meta.url);

const SOURCE_WEIGHTS = { rule: 1.0, reference: 0.8, project: 0.6 };

/**
 * 混合检索
 */
export async function hybridRetrieve(query, {
  taskType,
  topK = 5,
  ftsN = 15,
  vecN = 15
} = {}) {
  const startTime = Date.now();
  const queryStr = String(query);
  const queryLower = queryStr.toLowerCase();

  // 1. 任务路由
  const routeResult = taskType
    ? { task_type: taskType, ...router.getRoute(taskType) }
    : router.routeQuery(queryStr);
  const taskRoute = routeResult;

  // 2. 按任务路由的分类做硬过滤 + FTS 召回
  let ftsResults = [];

  // 如果有明确的任务分类，优先按分类范围过滤
  const categories = taskRoute.categories || [];
  const stages = taskRoute.stages || [];

  if (categories.length > 0) {
    // 先尝试按分类过滤 + 关键词模糊匹配
    const filtered = sqliteStore.filterChunks({ categories, stages, topN: ftsN });
    // 对过滤结果做简单的关键词评分
    ftsResults = filtered.map(row => {
      const text = (row.title + ' ' + row.text).toLowerCase();
      let score = 0;
      const terms = queryLower.split(/\s+/).filter(t => t.length > 0);
      for (const term of terms) {
        if (text.includes(term)) {
          score -= 1;  // FTS rank is negative = better match
        }
      }
      // sub-title match
      const titleText = row.title.toLowerCase();
      if (terms.some(t => titleText.includes(t))) {
        score -= 3;
      }
      return { row, fts_score: score < 0 ? score : -0.1 };
    }).filter(r => r.fts_score < 0);
  }

  // 如果分类过滤没结果，退回到全库 FTS
  if (ftsResults.length === 0) {
    ftsResults = sqliteStore.ftsSearch(queryStr, ftsN).map(r => ({
      row: r,
      fts_score: r.fts_score
    }));
  }

  const ftsMap = new Map();
  for (const r of ftsResults) {
    ftsMap.set(r.row.chunk_id, r);
  }

  // 3. 向量召回
  let vecResults = [];
  try {
    const queryVector = await embedText(queryStr);
    vecResults = vectorStore.vectorSearch(queryVector, vecN);
  } catch (err) {
    console.warn('[retriever] 向量检索失败:', err.message);
  }
  const vecMap = new Map();
  for (const r of vecResults) {
    vecMap.set(r.chunk_id, r.score);
  }

  // 4. 合并候选集
  const candidates = new Map();

  for (const [chunkId, info] of ftsMap) {
    candidates.set(chunkId, {
      chunk_id: chunkId,
      row: info.row,
      fts_score: info.fts_score,
      vec_score: vecMap.get(chunkId) || null
    });
  }

  for (const r of vecResults) {
    if (!candidates.has(r.chunk_id)) {
      const chunkData = sqliteStore.getChunkById(r.chunk_id);
      if (chunkData) {
        candidates.set(r.chunk_id, {
          chunk_id: r.chunk_id,
          row: chunkData,
          fts_score: null,
          vec_score: r.score
        });
      }
    }
  }

  // 5. 重排
  const reranked = Array.from(candidates.values()).map(c => {
    const score = rerank(c.row, queryLower, taskRoute, c.fts_score, c.vec_score);
    const reason = buildReason(c.row, c.fts_score, c.vec_score, taskRoute);
    return {
      chunk_id: c.chunk_id,
      title: c.row.title,
      score,
      reason,
      snippet: c.row.text.substring(0, 200) + (c.row.text.length > 200 ? '...' : ''),
      source_path: c.row.source_path,
      source_type: c.row.source_type,
      category: c.row.category,
      tags: c.row.tags || [],
      priority: c.row.priority
    };
  });

  reranked.sort((a, b) => b.score - a.score);
  const topResults = reranked.slice(0, topK);

  return {
    results: topResults,
    meta: {
      total_candidates: candidates.size,
      fts_count: ftsResults.length,
      vector_count: vecResults.length,
      elapsed_ms: Date.now() - startTime,
      task_type: routeResult.task_type,
      confidence: routeResult.confidence
    }
  };
}

/**
 * 重排
 */
function rerank(row, queryLower, taskRoute, ftsScore, vectorScore) {
  const ftsNorm = (typeof ftsScore === 'number' && ftsScore < 0)
    ? Math.min(1, Math.max(0, -ftsScore / 15))
    : 0;
  const vecNorm = (typeof vectorScore === 'number')
    ? Math.min(1, Math.max(0, vectorScore))
    : 0.2;

  let taskMatch = 0;
  if (taskRoute) {
    const cats = taskRoute.categories || [];
    if (cats.length > 0 && cats.includes(row.category)) taskMatch += 0.5;
    const stages = taskRoute.stages || [];
    if (row.stage && stages.some(s => row.stage.includes(s))) taskMatch += 0.3;

    const desc = (taskRoute.description || '').toLowerCase();
    const queryTerms = queryLower.split(/\s+/).filter(t => t.length > 1);
    const descTerms = desc.split(/[\s：:、，。]/);
    const overlap = queryTerms.filter(qt => descTerms.some(rt => rt.includes(qt)));
    taskMatch += Math.min(overlap.length * 0.05, 0.2);
  }
  taskMatch = Math.min(1, taskMatch);

  const priorityWeight = (6 - (row.priority || 3)) / 5;
  const sourceWeight = SOURCE_WEIGHTS[row.source_type] || 0.5;

  return Number((
    vecNorm * 0.35 +
    ftsNorm * 0.15 +
    taskMatch * 0.30 +
    priorityWeight * 0.12 +
    sourceWeight * 0.08
  ).toFixed(4));
}

/**
 * 构建 reason
 */
function buildReason(row, ftsScore, vectorScore, taskRoute) {
  const parts = [];
  if (typeof ftsScore === 'number' && ftsScore < 0) parts.push('关键词匹配');
  if (typeof vectorScore === 'number' && vectorScore > 0.3) parts.push('语义接近');
  if (taskRoute && (taskRoute.categories || []).includes(row.category)) parts.push('任务类型命中');
  if (row.priority <= 2) parts.push('高优先级');
  if (row.source_type === 'rule') parts.push('规则文档');
  if (parts.length === 0) parts.push('综合匹配');
  return parts.join(' + ');
}
